### PR TITLE
Support offset mapping alignment for fast tokenizers

### PR DESCRIPTION
**Pull request by adrianeboyd**
Switch to offset mapping-based alignment for fast tokenizers. With this change, slow vs. fast tokenizers will not give identical results with `spacy-transformers`.

Additional modifications:

* Update package setup for cython
* Update CI for compiled package

### CODE REVIEW

**Comment by adrianeboyd** [comment]
Concerns:

* naming of `_align` and new methods
* whether we (initially) want an automatic backoff to the old alignment if the new alignment fails in some cases
* further speed improvements
  * in particular I suspect this should be improved: https://github.com/explosion/spacy-transformers/pull/338/files#diff-f767b104e773bafeb21b7870c06043e8cebea5c4fce193b171b737192d589f62R90

File: spacy_transformers/_align.pyx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+
+cpdef get_span2wp_from_offset_mapping(span, wp_char_offsets):
+    # create a mapping of char indices to spacy token indices
+    cdef int span_idx = span[0].idx
+    cdef int span_i = span[0].i
+    cdef int char_idx, rel_token_i
+    # size is +1 so we don't have to check whether the text has a trailing space
+    char_to_sp_token = numpy.empty((len(span.text) + 1,), dtype="int32")
+    char_to_sp_token.fill(-1)
```
Comment (svlandeg) [inline spacy_transformers/_align.pyx:22]: Alternativelly, use `numpy.full`. It's definitely shorter, but not sure whether it's actually faster.

File: spacy_transformers/_align.pyx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    for token in span:
+        rel_token_i = token.i - span_i
+        for char_idx in range(
+                token.idx - span_idx,
+                token.idx - span_idx + len(token) + 1,
+        ):
+            char_to_sp_token[char_idx] = rel_token_i
+
+    # align all wordpiece tokens to one or more spacy token indicies
```
Comment (svlandeg) [inline spacy_transformers/_align.pyx:31]:
```suggestion
    # align all wordpiece tokens to one or more spacy token indices
```

File: spacy_transformers/_align.pyx [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# cython: infer_types=True, boundscheck=False
+from cython.operator cimport dereference as deref
+from cython.operator cimport preincrement as preinc
```
Comment (svlandeg) [inline spacy_transformers/_align.pyx:3]: I personally don't think that these aliases help with readability (especially `preinc` reads funny to me) but I don't care strongly if you prefer it like this.

File: spacy_transformers/_align.pyx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    ) nogil:
+    cdef int char_idx, start_idx, end_idx, token_i
+    cdef int wp_j = 0
+    cdef int alignment_size = alignment.size()
+    while wp_j < wp_char_offsets_length:
+        start_idx = wp_char_offsets[wp_j][0]
+        end_idx = wp_char_offsets[wp_j][1]
+        char_idx = start_idx
+        while char_idx < end_idx:
+            if char_idx >= 0 and char_idx < char_to_sp_token_length:
```
Comment (svlandeg) [inline spacy_transformers/_align.pyx:76]:
```suggestion
            if 0 <= char_idx < char_to_sp_token_length:
```

File: spacy_transformers/_align.pyx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                token_i = char_to_sp_token[char_idx]
+            else:
+                token_i = -1
+            if token_i >= 0 and token_i < alignment_size:
```
Comment (svlandeg) [inline spacy_transformers/_align.pyx:80]:
```suggestion
            if 0 <= token_i < alignment_size:
```

**Review by svlandeg — COMMENTED** [review]
Nice to finally get this properly working!

I suppose we'd want this as a minor release - should we create a `develop` branch for it? Or just remember to have the next one be 1.2.

File: spacy_transformers/_align.pyx (hunk 4 above)
Comment (adrianeboyd) [reply spacy_transformers/_align.pyx:76]: This was intentional to follow our usual `<` `>=` conventions, but I don't really care either way.

File: spacy_transformers/_align.pyx (hunk 3 above)
Comment (adrianeboyd) [reply spacy_transformers/_align.pyx:3]: This was copied from similar uses in `spacy`, I don't really care either way.

File: requirements.txt [hunk 6]
```diff
@@ -7,3 +7,4 @@ spacy-alignments>=0.7.2,<1.0.0
 # Testing
 pytest>=5.2.0
 pytest-cov>=2.7.0,<2.8.0
+cython>=0.25
```
[the remark this answers is not in the document]
Comment (adrianeboyd) [reply requirements.txt:10]: I think this is just an oversight.

File: spacy_transformers/_align.pyx (hunk 1 above)
Comment (adrianeboyd) [reply spacy_transformers/_align.pyx:22]: `numpy.full` says: New in version 1.20.0

File: spacy_transformers/tests/test_model_wrapper.py [hunk 7]
```diff
@@ -79,3 +79,12 @@ def test_model_predict(docs, trf_model):
     else:
         assert outputs.model_output.hidden_states is None
     assert isinstance(outputs, FullTransformerBatch)
+
+    # for a fast tokenizer check that all non-special wordpieces are aligned
+    # (which is not necessarily true for the slow tokenizers)
```
[the remark this answers is not in the document]
Comment (adrianeboyd) [reply spacy_transformers/tests/test_model_wrapper.py:84]: I didn't want to go to the trouble of implementing `offset_mapping` in the dummy tokenizer in the test suite. You could add separate tests for this in `test_alignment` with examples of full fast tokenizer output?

**Comment by adrianeboyd** [comment]
Yes, this would need to v1.2 because it will change the behavior of trained models.

File: spacy_transformers/align.py [hunk 8]
```diff
@@ -1,18 +1,20 @@
-import numpy
 from typing import cast, Dict, List, Tuple, Callable, Set, Optional
+import numpy
 from spacy_alignments.tokenizations import get_alignments
 from spacy.tokens import Span, Token
 from thinc.api import Ops
 from thinc.types import Ragged, Floats2d, Ints1d
 
+from ._align import get_span2wp_from_offset_mapping
```
[the remark this answers is not in the document]
Comment (adrianeboyd) [reply spacy_transformers/align.py:8]: All the names I can think of are clunky, maybe just the whole `align` module in cython? We'd just have to add the stub for the types.

File: spacy_transformers/_align.pyx (hunk 1 above)
Comment (adrianeboyd) [reply spacy_transformers/_align.pyx:22]: Ah, wait, I read that wrong and it was just for one kwarg. I just hadn't noticed it in our code before?

File: spacy_transformers/align.pyx [hunk 9]
```diff
@@ -79,22 +89,24 @@ def get_token_positions(spans: List[Span]) -> Dict[Token, int]:
     return token_positions
 
 
-def get_alignment_via_offset_mapping(spans: List[Span], token_data) -> Ragged:
-    # This function uses the offset mapping provided by Huggingface. I'm not
-    # sure whether there's a bug here but I'm getting weird errors.
+def get_alignment_via_offset_mapping(
+    spans: List[Span],
+    offset_mapping,
```
Comment (adrianeboyd) [inline spacy_transformers/align.pyx:94]: This should be typed. (Easier to test after #344.)

File: spacy_transformers/align.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (svlandeg) [reply spacy_transformers/align.py]: Yea, good call.

File: spacy_transformers/_align.pyx (hunk 3 above)
Comment (adrianeboyd) [reply spacy_transformers/_align.pyx]: I will leave this for now.

**Comment by adrianeboyd** [comment]
In terms of model performance and speed this implementation appears to be on par with the existing `spacy-alignments` alignment. In general I think it should able be to be little bit faster, my first guess (as mentioned above) is now here in the renamed version:

https://github.com/explosion/spacy-transformers/pull/338/files#diff-b326f7a5839c4589ffcac094a207251f14e6235d150b35a7ae07459776bcc19aR104

I'm trying to find a reasonable test case for where the actual alignments would make a difference, but most things I've tried so far were a wash. The main known cases are with roberta/gpt2 with the `bytes_to_unicode` mapping. At least users looking at the alignments would be less concerned about it?

**Review by shadeMe — COMMENTED** [review]
Looks good with just one minor comment. I don't see any immediate opportunities for optimization, but we should profile it as part of a deployed pipeline at some point.

**Pull request merged**